src/Microsoft.ML.TensorFlow/TensorFlow/Tensor.cs

```diff
@@ -446,6 +446,104 @@ internal static unsafe TFTensor CreateString(byte[] buffer)
+            }
+        }
+
+        internal static unsafe TFTensor CreateStringTensor(byte[][] buffer, TFShape shape)
+        {
+            if (buffer == null)
+                throw Contracts.ExceptEmpty(nameof(buffer));
+            //
+            // TF_STRING tensors are encoded with a table of 8-byte offsets followed by
+            // [offset1, offset2,...,offsetn, s1size, s1bytes, s2size, s2bytes,...,snsize,snbytes]
+            //
+            int size = 0;
+            foreach(var b in buffer)
+            {
+                size += (int)TF_StringEncodedSize((UIntPtr)b.Length);
+            }
+            int totalSize = size + buffer.Length * 8;
+            ulong offset = 0;
+            IntPtr handle = TF_AllocateTensor(TFDataType.String, shape.dims, shape.dims.Length, (UIntPtr)totalSize);
+
+            // Clear offset table
+            IntPtr pOffset = TF_TensorData(handle);
+            IntPtr dst = pOffset + buffer.Length * 8;
+            IntPtr dstLimit = pOffset + totalSize;
+            for (int i = 0; i < buffer.Length; i++)
+            {
+                Marshal.WriteInt64(pOffset, (long)offset);
+                using (var status = new TFStatus())
+                {
+                    fixed (byte* src = &buffer[i][0])
+                    {
+                        var written = TF_StringEncode(src, (UIntPtr)buffer[i].Length, (sbyte*)dst, (size_t)(dstLimit.ToInt64() - dst.ToInt64()), status.handle);
+                        var ok = status.StatusCode == TFCode.Ok;
+                        if (!ok)
+                            return null;
+                        pOffset += 8;
+                        dst = (IntPtr)(dst.ToInt64() + (Int64)written);
+                        offset += written.ToUInt64();
+                    }
+                }
+            }
+            return new TFTensor(handle);
+        }
+
+        internal static unsafe byte[][] DecodeStringTensor(TFTensor tensor)
+        {
+            if (tensor == null)
+                throw Contracts.ExceptEmpty(nameof(tensor));
+            //
+            // TF_STRING tensors are encoded with a table of 8-byte offsets followed by
+            // [offset1, offset2,...,offsetn, s1size, s1bytes, s2size, s2bytes,...,snsize,snbytes]
+            //
+            long size = 1;
+            foreach (var s in tensor.Shape)
+                size *= s;
+
+            var buffer = new byte[size][];
+            var src = TF_TensorData(tensor.handle);
+            var srcLen = (IntPtr)(src.ToInt64() + (long)tensor.TensorByteSize);
+            src += (int)(size * 8);
+            for (int i = 0; i < buffer.Length; i++)
+            {
+                using (var status = new TFStatus())
+                {
+                    IntPtr dst = IntPtr.Zero;
+                    UIntPtr dstLen = UIntPtr.Zero;
+                    var read = TF_StringDecode((sbyte*)src, (size_t)(srcLen.ToInt64() - src.ToInt64()), (sbyte**)&dst, &dstLen, status.handle);
```

Comment:
Is there a reason you are using `sbyte` here?

Looking at the TF header file:

[https://github.com/tensorflow/tensorflow/blob/3d704b754fd18f22a41336274bec9b5a85c3e4db/tensorflow/c/c_api.h#L328-L330](https://github.com/tensorflow/tensorflow/blob/3d704b754fd18f22a41336274bec9b5a85c3e4db/tensorflow/c/c_api.h#L328-L330)

```
TF_CAPI_EXPORT extern size_t TF_StringDecode(const char* src, size_t src_len,
                                             const char** dst, size_t* dst_len,
                                             TF_Status* status);
```

`dst` is a `char*`, just like `src`.  But in our `extern` functions are kind of messed up:

```
        // extern size_t TF_StringEncode (const char *src, size_t src_len, char *dst, size_t dst_len, TF_Status *status);
        [DllImport(NativeBinding.TensorFlowLibrary)]
        private static extern unsafe size_t TF_StringEncode(byte* src, size_t src_len, sbyte* dst, size_t dst_len, TF_Status status);

        // extern size_t TF_StringDecode (const char *src, size_t src_len, const char **dst, size_t *dst_len, TF_Status *status);
        [DllImport(NativeBinding.TensorFlowLibrary)]
        private static extern unsafe size_t TF_StringDecode(sbyte* src, size_t src_len, sbyte** dst, size_t* dst_len, TF_Status status);
```

We use `byte*` for TF_StringEncode's src. But sbyte else where.  I'd assume we should just use `byte*` everywhere. `byte` and `char` map together. Honestly, I'm not sure what would use `sbyte`...

Reply:
Yes, it needs to be fixed. This problem was brought in from TensorFlowSharp codebase where it is like this.
[https://github.com/migueldeicaza/TensorFlowSharp/blob/d3c2537f6e9aeea8119c7b05b847a8162ff8800d/TensorFlowSharp/Tensorflow.cs#L463](https://github.com/migueldeicaza/TensorFlowSharp/blob/d3c2537f6e9aeea8119c7b05b847a8162ff8800d/TensorFlowSharp/Tensorflow.cs#L463)

---
In reply to: [257056518](https://github.com/dotnet/machinelearning/pull/2546#discussion_r257056518) [](ancestors = 257056518)